src/AspNetCore/Src/inprocessapplication.cxx

```diff
@@ -107,6 +135,82 @@ http_read_request_bytes(
     _In_ IHttpContext* pHttpContext,
     _In_ CHAR* pvBuffer,
     _In_ DWORD cbBuffer,
+    _In_ DWORD* pDwBytesReceived,
+    _In_ BOOL* pfCompletionPending
+)
+{
+    IHttpRequest *pHttpRequest = (IHttpRequest*)pHttpContext->GetRequest();
+
+    BOOL fAsync = TRUE;
+
+    HRESULT hr = pHttpRequest->ReadEntityBody(
+        pvBuffer,
+        cbBuffer,
+        fAsync,
+        pDwBytesReceived,
+        pfCompletionPending);
+
+    if (hr == HRESULT_FROM_WIN32(ERROR_HANDLE_EOF))
+    {
+        // We reached the end of the data
+        hr = S_OK;
+    }
+
+    return hr;
+}
+
+EXTERN_C __MIDL_DECLSPEC_DLLEXPORT
+HRESULT
+http_write_response_bytes(
+    _In_ IHttpContext* pHttpContext,
+    _In_ HTTP_DATA_CHUNK* pDataChunks,
+    _In_ DWORD nChunks,
+    _In_ BOOL* pfCompletionExpected
+)
+{
+    IHttpResponse *pHttpResponse = (IHttpResponse*)pHttpContext->GetResponse();
+    BOOL fAsync = TRUE;
+    BOOL fMoreData = TRUE;
+    DWORD dwBytesSent = 0;
+
+    HRESULT hr = pHttpResponse->WriteEntityChunks(
+        pDataChunks,
+        nChunks,
+        fAsync,
+        fMoreData,
+        &dwBytesSent,
+        pfCompletionExpected);
+
+    return hr;
+}
+
+EXTERN_C __MIDL_DECLSPEC_DLLEXPORT
+HRESULT
+http_flush_response_bytes(
+    _In_ IHttpContext* pHttpContext,
```

Comment:
should we consider fAsync as input parameter

Reply:
Let's do that in a future PR.

Reply:
Actually, this may fix all of our problems @davidfowl . Can't we just check in managed if there is a currently running async operation? If there is, we just say fAsync is false?